PrestaShop/PrestaShop
# Update product.id_category_default after deleting categories

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Introduced new method in AbstractDeleteCategoryHandler which finds all products that has id_category_default amongst the recently deleted categories. For each of these products it finds parent category (parent of the one that was deleted, or fallbacks to a PS_HOME_CATEGORY) and assigns it, so that product doesn't end up being without default category. This method is called after handling the products which have no categories assigned at all (refactored a bit, but that ones logic doesn't change). Also it is moved out of a loop in bulk action, so it doesn't need to fetch the categories over and over again. Added behat tests (and refactored old tests which were really bad)
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yes (see bellow)
| Fixed ticket?     | Fixes #30219
| Related PRs       | 
| How to test?      | See bellow
| Possible impacts? | 

**deprecated** `PrestaShop\PrestaShop\Adapter\Category\CommandHandler\AbstractDeleteCategoryHandler::handleProductsUpdate`, Use `updateProductCategories` instead

**how to test**
Note - this scenario only applies if after category deletion there are other categories still left assigned with the product, but product ends up having no default category. In other case - If after category deletion the product is left without categories, then the "deletion mode" behavior applies as usually.

1. Create new categories "catA and catB".
2. Go to any product and assign those 2 categories and make catA default one. Make sure product now has only those 2 categories assigned.
3. Go to BO categories page and delete catA (no matter which deletion mode).
4. Go back to product page (v1 or v2 should both act same) and see that instead of the deleted "catA" the product default category is the parent category of the "catA" now
